resharper/resharper-unity/src/CSharp/Daemon/Stages/Highlightings/UnityHighlightingAbstractStage.cs

```diff
@@ -0,0 +1,115 @@
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.PerformanceCriticalCodeAnalysis.Analyzers;
+using JetBrains.ReSharper.Plugins.Unity.ProjectModel;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.Tree;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Highlightings
+{
+    public abstract class UnityHighlightingAbstractStage : CSharpDaemonStageBase
+    {
+        protected readonly IEnumerable<IUnityDeclarationHiglightingProvider> HiglightingProviders;
+        protected readonly UnityApi API;
+        private readonly UnitySolutionTracker mySolutionTracker;
+        protected readonly UnityHighlightingContributor UnityHighlightingContributor;
+
+        public UnityHighlightingAbstractStage(IEnumerable<IUnityDeclarationHiglightingProvider> higlightingProviders, UnityApi api, UnitySolutionTracker solutionTracker,
+            UnityHighlightingContributor unityHighlightingContributor, SolutionAnalysisService swa, PerformanceCriticalCodeCallGraphAnalyzer analyzer)
+        {
+            HiglightingProviders = higlightingProviders;
+            API = api;
+            mySolutionTracker = solutionTracker;
+            UnityHighlightingContributor = unityHighlightingContributor;
+        }
+        protected override IDaemonStageProcess CreateProcess(IDaemonProcess process, IContextBoundSettingsStore settings,
+            DaemonProcessKind processKind, ICSharpFile file)
+        {
+            if (!file.GetProject().IsUnityProject())
+                return null;
+            
+            return new UnityHighlightingProcess(process, file, HiglightingProviders, API, UnityHighlightingContributor, processKind);
+        }
+    }
+    
+    public class UnityHighlightingProcess : CSharpDaemonStageProcessBase
+    {
+        private readonly IEnumerable<IUnityDeclarationHiglightingProvider> myHiglightingProviders;
+        private readonly UnityApi myAPI;
+        private readonly UnityHighlightingContributor myUnityHighlightingContributor;
+        private readonly DaemonProcessKind myProcessKind;
+        private readonly ISet<IDeclaredElement> myMarkedDeclarations = new HashSet<IDeclaredElement>();
+        private readonly JetHashSet<IMethod> myEventFunctions;
+
+        public UnityHighlightingProcess([NotNull] IDaemonProcess process, [NotNull] ICSharpFile file,
+            IEnumerable<IUnityDeclarationHiglightingProvider> higlightingProviders, UnityApi api,
+            UnityHighlightingContributor unityHighlightingContributor, DaemonProcessKind processKind) : base(process, file)
+        {
+            myHiglightingProviders = higlightingProviders;
+            myAPI = api;
+            myUnityHighlightingContributor = unityHighlightingContributor;
+            myProcessKind = processKind;
+
+            myEventFunctions = DaemonProcess.CustomData.GetData(UnityEventFunctionAnalyzer.UnityEventFunctionNodeKey)
+                ?.ToJetHashSet();
```

Comment:
How do we know this has been populated by the analyser first?

Reply:
This stage executed after CSharpErrorStage